node/src/components/contract_runtime.rs

```diff
@@ -768,16 +772,28 @@ impl ContractRuntime {
         .await
         {
             Ok(block_and_execution_results) => block_and_execution_results,
-            Err(error) => return fatal!(effect_builder, "{}", error).await,
+            Err(error) => {
+                error!(%error, "failed to execute block");
+                return fatal!(effect_builder, "{}", error).await;
+            }
         };
 
         let new_execution_pre_state = ExecutionPreState::from_block_header(block.header());
-        debug!(
-            next_block_height = new_execution_pre_state.next_block_height,
-            "ContractRuntime: updating new_execution_pre_state",
-        );
-        *shared_pre_state.lock().unwrap() = new_execution_pre_state.clone();
-        debug!("ContractRuntime: updated new_execution_pre_state");
+        {
+            // The `shared_pre_state` could have been set to a block we just fully synced after
+            // doing a sync leap (via a call to `set_initial_state`).  We should not allow a block
+            // which completed execution just after this to set the `shared_pre_state` back to an
+            // earlier block height.
+            let mut shared_pre_state = shared_pre_state.lock().unwrap();
+            if shared_pre_state.next_block_height < new_execution_pre_state.next_block_height {
+                debug!(
+                    next_block_height = new_execution_pre_state.next_block_height,
+                    "ContractRuntime: updating new_execution_pre_state",
+                );
+                *shared_pre_state = new_execution_pre_state.clone();
+                debug!("ContractRuntime: updated new_execution_pre_state");
```

Comment:
Maybe one of these two similar `debug!` messages is redundant?
Also, wdyt about logging a message in case we detect the incorrect attempt to update execution state?

Reply:
I'm happy to remove one.  I'll remove the second one as it has less info.

Likewise, happy to add a debug message in an `else` arm.

Reply:
Fixed in [d24efc4](https://github.com/casper-network/casper-node/pull/3989/commits/d24efc48572d1701cbe4766ca8e71881a36a7bc8).